[WebGPU] GPUDevice bindGroup cache does not handle buffers with different offsets

mwyrzykowski opened the pull request:
#### 7605701f47be57dda96ead432207538a743d2302
<pre>
[WebGPU] GPUDevice bindGroup cache does not handle buffers with different offsets
<a href="https://bugs.webkit.org/show_bug.cgi?id=286407">https://bugs.webkit.org/show_bug.cgi?id=286407</a>
<a href="https://rdar.apple.com/143114463">rdar://143114463</a>

Reviewed by Cameron McCormack.

The GPUDevice GPUBindGroup cache incorrectly reused bind groups with
the same buffer but different offsets and sizes. Reuse should only be
applied to identical offsets and sizes otherwise and OOB read may occur.

* LayoutTests/fast/webgpu/nocrash/fuzz-286407-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-286407.html: Added.
Add regression test.

* Source/WebCore/Modules/WebGPU/GPUBindGroupEntry.h:
(WebCore::GPUBindGroupEntry::equalSizes):
(WebCore::GPUBindGroupEntry::equal):
Unlike GPUTexture, GPUSampler, and GPUExternalTextures, GPUBuffers
may specificy an offset into the buffer.

Canonical link: <a href="https://commits.webkit.org/289317@main">https://commits.webkit.org/289317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43b943e057acf3337c893f6acb70c743e5803a10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86299 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5853 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40636 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91220 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37110 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88352 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6107 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13890 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66840 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24638 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89305 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4640 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78178 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47157 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4453 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32495 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36207 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75011 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33365 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93151 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13506 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9819 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75688 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13717 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74060 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74870 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18427 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19068 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17456 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6286 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13528 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18826 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13284 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16725 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15070 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/Modules/WebGPU/GPUBindGroupEntry.h:
```diff
@@ -87,14 +87,18 @@ struct GPUBindGroupEntry {
             return false;
         });
     }
+    static bool equalSizes(const std::optional<GPUSize64>& a, const std::optional<GPUSize64>& b)
+    {
+        return (!a && !b) || (a && b && *a == *b);
+    }
```
According to https://en.cppreference.com/w/cpp/utility/optional/operator_cmp `std::optional::operator==` will do this check that the emptiness of both values is the same.

mwyrzykowski committed:
[WebGPU] GPUDevice bindGroup cache does not handle buffers with different offsets
https://bugs.webkit.org/show_bug.cgi?id=286407
rdar://143114463

Reviewed by Cameron McCormack.

The GPUDevice GPUBindGroup cache incorrectly reused bind groups with
the same buffer but different offsets and sizes. Reuse should only be
applied to identical offsets and sizes otherwise and OOB read may occur.

* LayoutTests/fast/webgpu/nocrash/fuzz-286407-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-286407.html: Added.
Add regression test.

* Source/WebCore/Modules/WebGPU/GPUBindGroupEntry.h:
(WebCore::GPUBindGroupEntry::equalSizes):
(WebCore::GPUBindGroupEntry::equal):
Unlike GPUTexture, GPUSampler, and GPUExternalTextures, GPUBuffers
may specificy an offset into the buffer.

Canonical link: https://commits.webkit.org/289317@main
